File: v1/ecmascript/food-chain/food-chain.js
function* range(begin, end) {
    for (let i = begin; i <= end; i++) {
        yield i;
    }
}

export default class Song {
    
    constructor() {
        this.numberOfVerses = 8;

        this.subjects = [
            "spider",
            "bird",
            "cat",
            "dog",
            "goat",
            "cow"
        ];

        this.followUps = [
            "It wriggled and jiggled and tickled inside her.",
            "How absurd to swallow a bird!",
            "Imagine that, to swallow a cat!",
            "What a hog, to swallow a dog!",
            "Just opened her throat and swallowed a goat!",
            "I don't know how she swallowed a cow!"
        ];

        this.history = [
            "I don't know how she swallowed a cow!",
            "She swallowed the cow to catch the goat.",
            "She swallowed the goat to catch the dog.",
            "She swallowed the dog to catch the cat.",
            "She swallowed the cat to catch the bird.",
            "She swallowed the bird to catch the spider that wriggled and jiggled and tickled inside her.",
            "She swallowed the spider to catch the fly.",
            "I don't know why she swallowed the fly. Perhaps she'll die."
        ];
    }
    
    verseBegin(number) {
        if (number == 1) {
            return "I know an old lady who swallowed a fly.";
        }

        if (number == 8) {
            return "I know an old lady who swallowed a horse.";
        }

        const subject = this.subjects[number - 2];
        const followUp = this.followUps[number - 2];
        return `I know an old lady who swallowed a ${subject}.\n${followUp}`;
    }

    verseEnd(number) {
        if (number == 8) {
            return "She's dead, of course!";
        }

        return this.history.slice(this.history.length - number, this.history.length - number + number).join("\n");
    }    
    
    verse(number) {
        return `${this.verseBegin(number)}\n${this.verseEnd(number)}\n`;
    }

    verses(begin, end) {
        const verses = [...range(begin, end)].map(number => this.verse(number) + "\n");
        return verses.join("");
    }
}

File: v1/haskell/food-chain/FoodChain.hs
module FoodChain (song) where

import qualified Data.Vector as V

song :: String
song = unlines $ map verse [0..verses -1]

verses :: Int
verses = 8

verse :: Int -> String
verse index = unlines $ (verseBegin index) ++ (verseEnd index)

verseBegin :: Int -> [String]
verseBegin 0 = ["I know an old lady who swallowed a fly."]
verseBegin 7 = ["I know an old lady who swallowed a horse."]
verseBegin index = ["I know an old lady who swallowed a " ++ subject ++ ".", followUp] where
    (subject, followUp) = subjects V.! (index - 1)

verseEnd :: Int -> [String]
verseEnd 7 = ["She's dead, of course!"]
verseEnd index = V.toList $ V.slice (V.length history - 1 - index) (index + 1) history

subjects :: V.Vector (String, String)
subjects = V.fromList [ ("spider", "It wriggled and jiggled and tickled inside her.")
                      , ("bird",   "How absurd to swallow a bird!")
                      , ("cat", "Imagine that, to swallow a cat!")
                      , ("dog", "What a hog, to swallow a dog!")
                      , ("goat", "Just opened her throat and swallowed a goat!")
                      , ("cow", "I don't know how she swallowed a cow!") ]

history :: V.Vector String
history = V.fromList [
    "I don't know how she swallowed a cow!"
  , "She swallowed the cow to catch the goat."
  , "She swallowed the goat to catch the dog."
  , "She swallowed the dog to catch the cat."
  , "She swallowed the cat to catch the bird."
  , "She swallowed the bird to catch the spider that wriggled and jiggled and tickled inside her."
  , "She swallowed the spider to catch the fly."
  , "I don't know why she swallowed the fly. Perhaps she'll die." ]